src/CodeGen_LLVM.cpp

```diff
@@ -512,32 +512,44 @@ llvm::Function *add_argv_wrapper(llvm::Module *m, llvm::Function *fn, const std:
 
 }
 
-void CodeGen_LLVM::compile_func(const LoweredFunc &f) {
-    const std::string &name = f.name;
-    const std::vector<Argument> &args = f.args;
-
+void CodeGen_LLVM::begin_func(const LoweredFunc &f) {
     // Deduce the types of the arguments to our function
-    vector<llvm::Type *> arg_types(args.size());
-    for (size_t i = 0; i < args.size(); i++) {
-        if (args[i].is_buffer()) {
+    vector<llvm::Type *> arg_types(f.args.size());
+    for (size_t i = 0; i < f.args.size(); i++) {
+        if (f.args[i].is_buffer()) {
             arg_types[i] = buffer_t_type->getPointerTo();
         } else {
-            arg_types[i] = llvm_type_of(args[i].type);
+            arg_types[i] = llvm_type_of(f.args[i].type);
         }
     }
 
     // Make our function
     FunctionType *func_t = FunctionType::get(i32, arg_types, false);
-    function = llvm::Function::Create(func_t, llvm_linkage(f.linkage), name, module.get());
+    function = llvm::Function::Create(func_t, llvm_linkage(f.linkage), f.name, module.get());
 
     // Mark the buffer args as no alias
-    for (size_t i = 0; i < args.size(); i++) {
-        if (args[i].is_buffer()) {
+    for (size_t i = 0; i < f.args.size(); i++) {
+        if (f.args[i].is_buffer()) {
             function->setDoesNotAlias(i+1);
         }
     }
 
-    debug(1) << "Generating llvm bitcode for function " << name << "...\n";
+    // If the Func is externally visible, also create the argv wrapper
```

Comment:
This seems a bit outside the meaning of "begin_func". How do you feel about factoring it into a third method? (As usual, I'm not happy about adding tons of methods, but that's the deal with refactoring. When you make smaller pieces, the pieces have to make sense on their own. I almost feel like this should be taken care of at the Module level. (And if the thought doesn't strike you as what you want to do, feel free to ignore the comment.)